components/provider_ppucd/src/structs.rs

```diff
@@ -0,0 +1,130 @@
+    pub const GR_EXT_V1: ResourceKey = resource_key!(uniset, "Gr_Ext", 1);
+    pub const GR_LINK_V1: ResourceKey = resource_key!(uniset, "Gr_Link", 1);
+    pub const HEX_V1: ResourceKey = resource_key!(uniset, "Hex", 1);
+    pub const HYPHEN_V1: ResourceKey = resource_key!(uniset, "Hyphen", 1);
+    pub const IDC_V1: ResourceKey = resource_key!(uniset, "IDC", 1);
+    pub const IDEO_V1: ResourceKey = resource_key!(uniset, "Ideo", 1);
+    pub const IDS_V1: ResourceKey = resource_key!(uniset, "IDS", 1);
+    pub const IDSB_V1: ResourceKey = resource_key!(uniset, "IDSB", 1);
+    pub const IDST_V1: ResourceKey = resource_key!(uniset, "IDST", 1);
+    pub const JOIN_C_V1: ResourceKey = resource_key!(uniset, "Join_C", 1);
+    pub const LOE_V1: ResourceKey = resource_key!(uniset, "LOE", 1);
+    pub const LOWER_V1: ResourceKey = resource_key!(uniset, "Lower", 1);
+    pub const MATH_V1: ResourceKey = resource_key!(uniset, "Math", 1);
+    pub const NCHAR_V1: ResourceKey = resource_key!(uniset, "NChar", 1);
+    pub const NFCINERT_V1: ResourceKey = resource_key!(uniset, "nfcinert", 1);
+    pub const NFDINERT_V1: ResourceKey = resource_key!(uniset, "nfdinert", 1);
+    pub const NFKCINERT_V1: ResourceKey = resource_key!(uniset, "nfkcinert", 1);
+    pub const NFKDINERT_V1: ResourceKey = resource_key!(uniset, "nfkdinert", 1);
+    pub const PAT_SYN_V1: ResourceKey = resource_key!(uniset, "Pat_Syn", 1);
+    pub const PAT_WS_V1: ResourceKey = resource_key!(uniset, "Pat_WS", 1);
+    pub const PCM_V1: ResourceKey = resource_key!(uniset, "PCM", 1);
+    pub const PRINT_V1: ResourceKey = resource_key!(uniset, "print", 1);
+    pub const QMARK_V1: ResourceKey = resource_key!(uniset, "QMark", 1);
+    pub const RADICAL_V1: ResourceKey = resource_key!(uniset, "Radical", 1);
+    pub const RI_V1: ResourceKey = resource_key!(uniset, "RI", 1);
+    pub const SD_V1: ResourceKey = resource_key!(uniset, "SD", 1);
+    pub const SEGSTART_V1: ResourceKey = resource_key!(uniset, "segstart", 1);
+    pub const SENSITIVE_V1: ResourceKey = resource_key!(uniset, "Sensitive", 1);
+    pub const STERM_V1: ResourceKey = resource_key!(uniset, "STerm", 1);
+    pub const TERM_V1: ResourceKey = resource_key!(uniset, "Term", 1);
+    pub const UIDEO_V1: ResourceKey = resource_key!(uniset, "UIdeo", 1);
+    pub const UPPER_V1: ResourceKey = resource_key!(uniset, "Upper", 1);
+    pub const VS_V1: ResourceKey = resource_key!(uniset, "VS", 1);
+    pub const WSPACE_V1: ResourceKey = resource_key!(uniset, "WSpace", 1);
+    pub const XDIGIT_V1: ResourceKey = resource_key!(uniset, "xdigit", 1);
+    pub const XIDC_V1: ResourceKey = resource_key!(uniset, "XIDC", 1);
+    pub const XIDS_V1: ResourceKey = resource_key!(uniset, "XIDS", 1);
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct UnicodeProperties<'s> {
+    pub props: Vec<UnicodeProperty<'s>>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct UnicodeProperty<'s> {
+    pub name: &'s str,
```

Comment:
Actually, I should have caught this yesterday, but this should be `Cow<'s, str>`.  Most everything works the same, but you'll need to wrap strings you save here with `Cow::Borrowed()`.

Reply:
Done.